### PR TITLE
Add PruneRewardAccountingTrigger to delete old reward-accounting data

### LINKED ISSUE #6136: Prune TBAR stores that hold intermediate date
That is the tables backing these two stores:
- https://github.com/canton-network/splice/blob/d5d1108aa222955b4fac192feb6881164911e9cb/apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanAppRewardsStore.scala#L15
- https://github.com/canton-network/splice/blob/d5d1108aa222955b4fac192feb6881164911e9cb/apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala#L31

The goal is that the performance of the reward computation is independent of the time it has been running. 

Note that the pruning of the `AppActivityStore` will be done separately, as part of the work be the Splice Maintainers to store the verdicts in a bulk format outside of the Scan DB.

### PR DESCRIPTION
**Pull request by dfordivam**
Fixes #6136

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If an upgrade test is required, comment `/upgrade_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a logical synchronizer upgrade test is required (from canton-3.5), comment `/lsu_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/canton-network/splice/blob/main/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/canton-network/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala [hunk 1]
```diff
@@ -685,6 +690,95 @@ class TrafficBasedRewardsSvAppTimeBasedIntegrationTest
     )
   }
 
+  private def advanceRoundsAndConfirmAllPriorRoundsPruned()(implicit
```
Comment (dfordivam) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala:790]: This test is flaky at the moment, so I am investigating to make it more robust

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanAppRewardsStore.scala [hunk 2]
```diff
@@ -1020,6 +1030,49 @@ class DbScanAppRewardsStore(
             )
     """.asUpdate
 
+  /** Deletes all reward-accounting data for a single round across all six
+    * tables, in a single transaction.
+    */
+  def deleteRewardAccountingDataForRound(
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanAppRewardsStore.scala:1038]: Indexes being used here for reference

```
    "app_reward_batch_hashes_pkey" PRIMARY KEY, btree (history_id, round_number, batch_level, party_seq_num_begin_incl)
    "app_reward_root_hashes_pkey" PRIMARY KEY, btree (history_id, round_number)
    "app_reward_party_totals_pkey" PRIMARY KEY, btree (history_id, round_number, app_provider_party_seq_num)
    "app_reward_round_totals_pkey" PRIMARY KEY, btree (history_id, round_number)
    "app_activity_party_totals_pkey" PRIMARY KEY, btree (history_id, round_number, app_provider_party)
    "app_activity_round_totals_pkey" PRIMARY KEY, btree (history_id, round_number)
```

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala [hunk 3]
```diff
@@ -270,6 +270,85 @@ class DbScanRewardsReferenceStore(
...
+                      and active.round <= archived.round
+                  )
+           """.as[Option[Long]].head,
+          "lookupLowestPrunableArchivedRewardRound",
+        )
+      )
+    }
+
+  // Returns number of rows deleted
+  def pruneArchivedDataForRound(
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala]: This will always be a seq scan

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+       *
+       * Also here it is assumed that ScanAppRewardsStore only has the data for
+       * rounds for which we have data in ScanRewardsReferenceStore. Which is a
+       * fair assumption as the reward calculations require the data to be
+       * present in ScanRewardsReferenceStore for that round.
+       */
+      summary <- appRewardsStore.deleteRewardAccountingDataForRound(task.roundNumber)
+      deletedArchiveRows <- rewardsReferenceStore.pruneArchivedDataForRound(task.roundNumber)
+    } yield TaskSuccess(
+      s"Pruned reward accounting data for round ${task.roundNumber}: " +
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala]: Example log

    "Completed processing with outcome: Pruned reward accounting data for round 11: removed 5 archived rows; and RewardAccountingPruneSummary(2,1,2,1,1,1)."

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+              logger.warn(
+                s"This should never happen, could not resolve archived_at for round $roundNumber."
+              )
+              Seq.empty
+            case Some(uptoInclusive) =>
+              if (verdictStore.lastIngestedRecordTime.exists(_ >= uptoInclusive)) {
+                Seq(Task(roundNumber))
+              } else {
+                logger.debug(
+                  s"Skipping pruning as the verdict ingestion (last ingested record time: ${verdictStore.lastIngestedRecordTime}) " +
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala]: Example log

    "Skipping pruning as the verdict ingestion (last ingested record time: Some(1970-01-01T02:02:00.000131Z)) has not yet caught up to round 11's archivedAt (1970-01-01T02:02:00.000222Z)"

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                    } yield (activeRoundO, archivedAtNextO) match {
+                      case (Some(activeRound), Some(archivedAtNext)) =>
+                        if (
+                          CantonTimestamp
+                            .assertFromInstant(activeRound.payload.opensAt) > archivedAtNext
+                        ) {
+                          Seq(Task(roundNumber))
+                        } else {
+                          logger.debug(
+                            s"Skipping pruning of round $roundNumber as the ingestion's currently active round " +
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala]: This is the log, though this may be noisy and perhaps can be removed

   "Skipping pruning of round 12 as the ingestion's currently active round 15 opened at 1970-01-01T17:47:20Z, which is not yet past round 13's archivedAt (1970-01-01T17:47:30.000228Z)."

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala (hunk 1 above)
Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala:790]: Fixed the flakyness, needed more time after `pauseScanVerdictIngestionWithin` block

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala [hunk 7]
```diff
@@ -270,6 +270,85 @@ class DbScanRewardsReferenceStore(
       case _ => Future.successful(None)
     }
 
+  def lookupArchivedAtForOpenMiningRound(
```
Comment (adetokunbo) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala]: nit: use override here, to be consistent with `lookupLowestPrunableArchiveRewardRound` below
Also for `pruneArchiveDataForRound`

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala [hunk 8]
```diff
@@ -685,6 +690,108 @@ class TrafficBasedRewardsSvAppTimeBasedIntegrationTest
...
+    }
+
+    clue(s"sv2 eventually prunes data once verdict ingestion resumes") {
+      // This can occasionally take longer than the default 20s eventually window
+      confirmFullyPruned(
+        sv2Db,
+        sv2HistoryId,
+        sv2RewardsRefStore,
+        newLowestOpen,
+        timeUntilSuccess = 90.seconds,
```
Comment (adetokunbo) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala]: The comment above suggests that it can complete in 20s, but is sometimes much longer; is it possible to also document why there is such a large variance as a means of justifying the 90s ?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import org.lfdecentralizedtrust.splice.scan.store.db.DbScanVerdictStore
+import org.lfdecentralizedtrust.splice.store.UpdateHistory
+import com.digitalasset.canton.data.CantonTimestamp
+import com.digitalasset.canton.logging.pretty.{Pretty, PrettyPrinting}
+import com.digitalasset.canton.tracing.TraceContext
+import io.opentelemetry.api.trace.Tracer
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Periodically deletes reward-accounting data that is no longer needed, one
```
Comment (adetokunbo) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:27]: Design question:  it's not mentioned in #6136, so asking here: is it confirmed that no retention window is necessary?  The data in these tables is useful for debugging.  IUC, this prunes them aggressively - should there be a configurable capability to retain them for a bit longer?

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala [hunk 10]
```diff
@@ -685,6 +690,108 @@ class TrafficBasedRewardsSvAppTimeBasedIntegrationTest
...
+      (1 to 3).foreach(_ => advanceRoundsToNextRoundOpening)
+
+      clue(s"sv1 prunes rounds below $newLowestOpen") {
+        confirmFullyPruned(sv1Db, sv1HistoryId, sv1RewardsRefStore, newLowestOpen)
+      }
+
+      clue(
+        s"sv2 does not prune rounds below $newLowestOpen while its verdict ingestion is paused"
+      ) {
+        hasUnprunedArchiveDataForRound(sv2RewardsRefStore, newLowestOpen - 1) shouldBe true
```
Comment (rautenrieth-da) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala]: Should we explicitly trigger one iteration of `PruneRewardAccountingTrigger` here? This might mean that we have to pause the trigger for the entire duration of the test and control them manually. Although unlikely, it could be that sv2 has unpruned data simply because the trigger didn't finish yet.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala (hunk 8 above)
Comment (rautenrieth-da) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala]: Also interested in knowing why it takes so long. Is it verdict ingestion taking so long to catch up, or the pruning trigger?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+          verdictStore.lastIngestedRecordTime match {
+            case None =>
+              logger.debug("Skipping pruning as no verdict has been ingested yet.")
+              Future.successful(Seq.empty)
+            case Some(lastIngestedRecordTime) =>
+              rewardsReferenceStore
+                .lookupActiveOpenMiningRounds(Seq(lastIngestedRecordTime))
+                .flatMap(_.get(lastIngestedRecordTime) match {
+                  case None =>
+                    logger.warn(
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala]: Tests are currently hitting this line. Could it be that the verdict ingestion is ahead of the reference store ingestion so that `lookupActiveOpenMiningRounds(lastIngestedRecordTime)` doesn't have an answer yet?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala [hunk 12]
```diff
@@ -94,6 +94,30 @@ trait ScanRewardsReferenceStore extends AppStore {
       tc: TraceContext
   ): Future[Seq[Contract[CalculateRewardsV2.ContractId, CalculateRewardsV2]]]
 
+  def lookupArchivedAtForOpenMiningRound(
+      roundNumber: Long
+  )(implicit tc: TraceContext): Future[Option[CantonTimestamp]]
+
+  /** The lowest round number that may be safe to prune
+    * The criterion is based on the prescence of `OpenMiningRound`,
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala]:
```suggestion
    * The criterion is based on the presence of `OpenMiningRound`,
```

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+  *
+  *   - Archived contracts from `ScanRewardsReferenceStore` archive table
+  *     which archived <= the archival time of the OpenMiningRound
+  *
+  * A round only becomes eligible for pruning once none of its `OpenMiningRound`,
+  * `CalculateRewardsV2`, or `ProcessRewardsV2` contracts -- nor those of any
+  * lower round -- remain active. AND the verdict ingestion has moved past the
+  * `record_time` of OpenMiningRound's archival.
+  */
+class PruneRewardAccountingTrigger(
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:47]: Can you please add some basic metrics to this trigger? Maybe a `Gauge` with the latest round eligible for pruning, or some meters for the values in `RewardAccountingPruneSummary`?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala [hunk 14]
```diff
@@ -270,6 +270,85 @@ class DbScanRewardsReferenceStore(
...
+      // even after the pruning.
+      val pqn = PackageQualifiedName.fromJavaCodegenCompanion(OpenMiningRound.COMPANION)
+      futureUnlessShutdownToFuture(
+        storage.query(
+          sql"""select min(archived.round)
+                from #${ScanRewardsReferenceTables.archiveTableName} archived
+                where archived.store_id = $storeId and archived.migration_id = $migrationId
+                  and archived.package_name = ${pqn.packageName}
+                  and archived.template_id_qualified_name = ${pqn.qualifiedName}
+                  and archived.round is not null
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala]: Can the round ever be null for OpenMingRound contracts?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala [hunk 15]
```diff
@@ -270,6 +270,85 @@ class DbScanRewardsReferenceStore(
...
+      val storeId = multiDomainAcsStore.acsStoreId
+      val migrationId = multiDomainAcsStore.domainMigrationId
+      // It is important to limit the query for the round to OpenMiningRound
+      // as we prune only the contracts <= OpenMiningRound's archived_at
+      // and the db will likely still have other contracts for this round
+      // even after the pruning.
+      val pqn = PackageQualifiedName.fromJavaCodegenCompanion(OpenMiningRound.COMPANION)
+      futureUnlessShutdownToFuture(
+        storage.query(
+          sql"""select min(archived.round)
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala]: This relies on the fact that archiving an OpenMiningRound also creates a CalculateRewardsV2 contract in the same transaction, right?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+          }
+      }
+
+  override def completeTask(task: Task)(implicit tc: TraceContext): Future[TaskOutcome] =
+    for {
+      /* The two deletes happen on different stores, and the deletes are being
+       * done in independent transactions.
+       * Because both are idempotent operations and re-running either for the
+       * same or an older round is a no-op.
+       * If pruning of either store fails, the task will be retried.
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:100]: We also need to consider the case where the task fails in between the two pruning operations, and then all retries fail because of intermittent database connection issues.

In that case the task will be abandoned, and a new task eventually created. The logic for creating the new task needs to work even if exactly one of the two independent pruning transactions completed. It looks like this works because we first delete the reward accounting data, and the result of `retrieveTasks` only depends on the TCS store which is pruned second. I would mention this in the comment though.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala [hunk 17]
```diff
@@ -94,6 +94,30 @@ trait ScanRewardsReferenceStore extends AppStore {
...
+
+  /** Deletes all rows from the archive table with `archived_at <=` the record_time at
+    * which round's `OpenMiningRound` contract was archived.
+    *
+    * Here we don't have a lower bound on the `archived_at`, so this API should
+    * be used with care, preferably ensuring that the deletion happens one round at a time.
+    *
+    * Returns number of rows deleted.
+    */
+  def pruneArchivedDataForRound(
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala]:
```suggestion
  def pruneArchivedDataBeforeRound(
```
or similar, "for" sounds like it would only delete data for that particular round

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 9 above)
Comment (rautenrieth-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:27]: Good point! I think a configurable retention window with a default of 1 week (or at least a day) would be very useful for debugging. In case of a high severity issue like a root hash mismatch we might want to inspect the state of the database, and it might take a while to get a data dump from all SVs.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+    override val tracer: Tracer,
+) extends PollingParallelTaskExecutionTrigger[Task] {
+
+  override protected def context: TriggerContext = triggerContext.copy(
+    config = triggerContext.config.copy(
+      parallelism = 1
+    )
+  )
+
+  override def retrieveTasks()(implicit tc: TraceContext): Future[Seq[Task]] =
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:72]: The logic in this method is quite complex. Can we unit test it? If you want to avoid setting up the stores with real contracts, you could factor out the implementation into an independent function and then call it with test values, like we did with `AcsSnapshotTrigger.retrieveTaskForCurrentMigration`.

**Review by rautenrieth-da — COMMENTED** [review]
Thanks! Wasn't easy to review, but looking good in general. Left some minor comments.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala (hunk 8 above)
Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala]: It is verdict ingestion, the cause is some grpc stream issue where it errors out if not consumed, and it need retry by the service.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala (hunk 14 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbScanRewardsReferenceStore.scala]: not really, I was probably trying to make use of the index, but this can be removed as we have another one to cover this.

```
    "scan_rewards_reference_store_archived_round" btree (store_id, migration_id, round) WHERE round IS NOT NULL
    "scan_rewards_reference_store_archived_temporal" btree (store_id, migration_id, package_name, template_id_qualified_name, archived_at) INCLUDE (created_at)
```

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 18 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:72]: yeah now that we even have retention window this did get complex, and even testing it in a unit test was tricky; mostly cos its hard to judge whether the testing is good enough and if the test makes sense.

And I did notice a bug with the in-memory var cache of TcsStore ingestion-start, which is also fixed now.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala (hunk 17 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala]: renamed to `pruneArchivedUpToRound` as we actually delete the data which was for in store when the round was active.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 16 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:100]: Done

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 11 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala]: yeah this was a mistake, changed to a debug. Its a bootstrap related issue, as `lookupActiveOpenMiningRounds` doesn't work for rounds 0, 1 on non initial SVs.

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala (hunk 10 above)
Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsSvAppTimeBasedIntegrationTest.scala]: Done, now being run manually

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 9 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:27]: Done, added `rewardAccountingRetentionPeriod` config with default of 1 week

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 13 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:47]: ok, added metrics as described

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/metrics/RewardAccountingPruningMetrics.scala [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import org.lfdecentralizedtrust.splice.scan.store.db.DbScanAppRewardsStore.RewardAccountingPruneSummary
+
+class RewardAccountingPruningMetrics(metricsFactory: LabeledMetricsFactory)(implicit
+    metricsContext: MetricsContext
+) extends AutoCloseable {
+  private val prefix: MetricName =
+    SpliceMetrics.MetricsPrefix :+ "scan" :+ "reward_accounting_pruning"
+
+  // In normal operation this advances by one round per tick, so it stalling is
+  // the signal that pruning stopped making progress.
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/metrics/RewardAccountingPruningMetrics.scala:19]: Opened https://github.com/canton-network/splice/issues/7186 for implementing the alert.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala [hunk 20]
```diff
@@ -99,6 +95,91 @@ trait ScanRewardsReferenceStore extends AppStore {
...
+        // archived_at of round roundNumber + 1.
+        for {
+          activeRoundO <- lookupActiveOpenMiningRounds(Seq(lastIngestedRecordTime))
+            .map(_.get(lastIngestedRecordTime))
+          archivedAtNextO <- lookupArchivedAtForOpenMiningRound(roundNumber + 1)
+        } yield (activeRoundO, archivedAtNextO) match {
+          case (Some((activeRoundNumber, activeRoundOpensAt)), Some(archivedAtNext)) =>
+            val retainUntil = now.minus(retentionPeriod.asJava)
+            if (activeRoundOpensAt <= archivedAtNext) {
+              logger.debug(
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/ScanRewardsReferenceStore.scala]: In practice (with a retention period of 7 days), we should never run into situations where a round outside of the retention period has to be skipped because the pruned round is too close to the currently active round, no? Can we maybe swap the two branches so we get the "is still within the retention period" log instead of this complicated one?

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala [hunk 21]
```diff
@@ -426,6 +421,408 @@ class DbScanRewardsReferenceStoreTest
...
+
+          _ <- sync1.archive(pr3, recordTime = ts(230).toInstant)(store.multiDomainAcsStore)
+          afterPr3Archived <- store.lookupLowestPrunableArchivedRewardRound()
+          _ = afterPr3Archived shouldBe Some(3L)
+        } yield succeed
+      }
+
+    }
+
+    "lookupPrunableRewardRound" should {
```
Comment (rautenrieth-da) [inline apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala:643]: Is this missing testing the `Skipping pruning of round $roundNumber as round ${roundNumber + 1} has not archived yet.` branch?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    triggerContext: TriggerContext,
+)(implicit
+    override val ec: ExecutionContext,
+    mat: Materializer,
+    override val tracer: Tracer,
+) extends PollingParallelTaskExecutionTrigger[Task] {
+
+  override protected def context: TriggerContext = triggerContext.copy(
+    config = triggerContext.config.copy(
+      parallelism = 1
```
Comment (rautenrieth-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:62]: We could probably also tweak the polling interval. Running the trigger every 30sec is wasteful if it can only prune data once every 10min. I don't think we need to do this now though.

**Review by rautenrieth-da — APPROVED** [review]
> TODO: add release notes

This is still missing, right? Otherwise the PR looks good, thanks!

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala (hunk 22 above)
Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/PruneRewardAccountingTrigger.scala:62]: yeah, a 2 min interval may be good. To clear the backlog at 5x pace should be good enough.

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala (hunk 21 above)
Comment (dfordivam) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala:643]: Actually adding a test for this branch turned out to be a little tricky, and not that useful.
The reason being the "ingestion start" of the store need to be before the openAt of round, otherwise we will hit the `case (None, _)` branch.
And I will have to fabricate a store state with an archived contract like `FeaturedAppRight` to exist in the archived table before the `openAt` of the currently open round, while also ensuring that the archived contract of the previous round is not in archive table. This scenario can be created in test, but can't happen if we follow the logic of round advancement and pruning.
Do you see a value in adding this test? or I should instead change the message to "this can never happen"

Comment (rautenrieth-da) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala:643]: I see. No need to force it in a test then. Consider bumping the log message to a warning and adding a "This should never happen" to the text.

Comment (dfordivam) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/DbScanRewardsReferenceStoreTest.scala:643]: Ok converted this to a warn, and also added a store bootstrap / ingestion bootstrap test for the other `case (None, _)` branch. This test also acts as a demonstration to assert that the warn wont hit.
